### DISCUSSION TITLE
Generate client based on schema only from graphql-kotlin-maven-plugin -> generate-sdl

### DISCUSSION THREAD

**Discussion by AlehGalo** [ExpediaGroup/graphql-kotlin#1556] [Q&A] [answered]
Hi. 

We are baking a startup and came across usage of this lib.
The application is built on top of micro-services. We have 2 services running with GraphQL-kotlin.
The 3rd service is a middle service and we have to generate client for those 2 services.

I wonder if it is possible to generate client just using only schema.graphql generated from 2 services with the usage of graphql-kotlin-maven-plugin -> generate-sdl ? Or how can I automatically generate query files based on sources that can be used for client generation ?

While using generate-client goal it requires query files. I believe that query is already specified in schema.graphql

My schema file

```

schema {
  query: Query
}

"Marks the field, argument, input field or enum value as deprecated"
directive @deprecated(
    "The reason for the deprecation"
    reason: String = "No longer supported"
  ) on FIELD_DEFINITION | ARGUMENT_DEFINITION | ENUM_VALUE | INPUT_FIELD_DEFINITION

"Directs the executor to include this field or fragment only when the `if` argument is true"
directive @include(
    "Included when true."
    if: Boolean!
  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Directs the executor to skip this field or fragment when the `if` argument is true."
directive @skip(
    "Skipped when true."
    if: Boolean!
  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Exposes a URL that specifies the behaviour of this scalar."
directive @specifiedBy(
    "The URL that specifies the behaviour of this scalar."
    url: String!
  ) on SCALAR

interface AbstractPositionable {
  h: Int!
  w: Int!
  x: Int!
  y: Int!
}

interface PrimaryKeyUUID {
  uuid: UUID!
}

interface SequentialUuid implements PrimaryKeyUUID {
  prevUuid: UUID!
  uuid: UUID!
}

type BoardDto implements PrimaryKeyUUID {
  archivedDateTime: LocalDateTime!
  description: String!
  insertDate: LocalDateTime!
  name: String!
  plannerUuid: UUID!
  sections: [SectionDto!]!
  userUuid: UUID!
  uuid: UUID!
  widgets: [WidgetDto!]!
}

type ItemDto {
  insertDate: LocalDateTime!
  refUuid: UUID!
  type: Type!
  uuid: UUID!
}

type SectionDto implements PrimaryKeyUUID & SequentialUuid {
  boardUuid: UUID!
  items: [WidgetDto!]!
  name: String!
  type: SectionType!
  uuid: UUID!
}

type WidgetDto implements AbstractPositionable & PrimaryKeyUUID {
  copyRefUuid: UUID!
  h: Int!
  insertDate: LocalDateTime!
  isSingle: Boolean!
  items: [ItemDto!]!
  name: String!
  plannerUuid: UUID!
  sectionUuid: UUID!
  type: Type!
  uuid: UUID!
  w: Int!
  x: Int!
  y: Int!
}

enum SectionType {
  GRID
  PAGE
}

enum Type {
  EMAIL
  TEXT
  WORK_ITEM
}

"A type representing a formatted java.util.LocalDateTime"
scalar LocalDateTime

"A type representing a formatted java.util.UUID"
scalar UUID

```

Like for Open Api I can use openapi-generator-maven-plugin that will generate all needed client code based on 1 open api json specification.

With respect.
Thanks in advance.

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 

There is a major difference between OpenAPI and GraphQL in that in the latter clients decide which fields they want to request (vs server dictating the full response in REST). GraphQL schema file defines the shape of your API and clients need to explicitly provide queries that define which data they want. Thats why you need both schema and query files in order to generate the client. 

There is no wildcard support in GraphQL and queries have to be explicit (see [this issue](https://github.com/graphql/graphql-spec/issues/942)), i.e. there is no `SELECT * FROM schema` SQL equivalent. If your clients always request full schema then I am unsure whether you get much out of GraphQL and OpenAPI might be potentially better suited for your use cases.

For more details see https://graphql.org/learn/.

**Reply by AlehGalo** [reply]
Hi thank you for your reply. I really appreciate it.

Went through it multiple times. And a question in my head is still unaswered.

I was able to make an experiment with DSG.

Schema

```
schema {
  query: Query
}

type Query {
  getBoard(id: UUID!): BoardDto!
  getBoardPlain(id: UUID!): BoardDto!
}


"Marks the field, argument, input field or enum value as deprecated"
directive @deprecated(
    "The reason for the deprecation"
    reason: String = "No longer supported"
  ) on FIELD_DEFINITION | ARGUMENT_DEFINITION | ENUM_VALUE | INPUT_FIELD_DEFINITION

"Directs the executor to include this field or fragment only when the `if` argument is true"
directive @include(
    "Included when true."
    if: Boolean!
  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Directs the executor to skip this field or fragment when the `if` argument is true."
directive @skip(
    "Skipped when true."
    if: Boolean!
  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Exposes a URL that specifies the behaviour of this scalar."
directive @specifiedBy(
    "The URL that specifies the behaviour of this scalar."
    url: String!
  ) on SCALAR

interface AbstractPositionable {
  h: Int!
  w: Int!
  x: Int!
  y: Int!
}

interface PrimaryKeyUUID {
  uuid: UUID!
}

interface SequentialUuid implements PrimaryKeyUUID {
  prevUuid: UUID!
  uuid: UUID!
}

type BoardDto implements PrimaryKeyUUID {
  archivedDateTime: LocalDateTime!
  description: String!
  insertDate: LocalDateTime!
  name: String!
  plannerUuid: UUID!
  sections: [SectionDto!]!
  userUuid: UUID!
  uuid: UUID!
  widgets: [WidgetDto!]!
}

type ItemDto {
  insertDate: LocalDateTime!
  refUuid: UUID!
  type: Type!
  uuid: UUID!
}

type SectionDto implements PrimaryKeyUUID & SequentialUuid {
  boardUuid: UUID!
  content: String!
  insertDate: LocalDateTime!
  items: [WidgetDto!]!
  name: String!
  prevUuid: UUID!
  type: SectionType!
  uuid: UUID!
}

type WidgetDto implements AbstractPositionable & PrimaryKeyUUID {
  copyRefUuid: UUID!
  h: Int!
  insertDate: LocalDateTime!
  isSingle: Boolean!
  items: [ItemDto!]!
  name: String!
  plannerUuid: UUID!
  sectionUuid: UUID!
  type: Type!
  uuid: UUID!
  w: Int!
  x: Int!
  y: Int!
}

enum SectionType {
  CANVAS
  GRID
  PAGE
}

enum Type {
  ASSEMBLED
  CALENDAR
  CUSTOM_APP
  EMAIL
  EMAIL_FOLDER
  EMBEDDED_LINK
  FILE
  MEETING
  MEETING_AGENDA
  MEETING_MINUTES
  NOTE
  PLANNER
  STORY_MAP
  SWIMLANE
  TEXT
  WEB_PASTE
  WORKSTREAM
  WORK_ITEM
}

"A type representing a formatted java.util.LocalDateTime"
scalar LocalDateTime

"A type representing a formatted java.util.UUID"
scalar UUID

```

And for the schema above DSG generated all stuff around the GQL client.

<img width="622" alt="Screenshot 2022-09-18 at 21 43 26" src="https://user-images.githubusercontent.com/3691373/190923452-5031bd0e-2317-4625-9ef7-a13c06bfee02.png">

Is GraphQL-kotlin plugin able to make the same without creation of any queries from human side ?

In other words am I able to push just only 1 grapqhl schema and get fully working client library for the service ?

As a result I can build the query by DSG:

<img width="957" alt="Screenshot 2022-09-18 at 21 54 41" src="https://user-images.githubusercontent.com/3691373/190923827-3697c158-c320-468e-a914-affc0447e7bc.png">

Thanks in advance.
Appreciate your time and effort.

**Reply by dariuszkuc** [reply]
I am not familiar with DGS client generation so cannot say for certain but it would appear to me that they are auto generating multiple (all?) code permutations and allow end users to manually build queries from the generated code. IMHO this is not really scaleable approach. 

Client code generation is generally driven by the queries (e.g. why you would want to generate "dead" code that is not going to be used?) and thats the approach that `graphql-kotlin` (and many others) follow. So to reiterate -> you will need both schema file and query file(s) to generate your client code using `graphql-kotlin`.
